src/main/groovy/org/assertj/generator/gradle/tasks/config/Templates.groovy

```diff
@@ -33,248 +37,261 @@ import static org.assertj.assertions.generator.Template.Type.*
      */
-    ListProperty<File> templateFiles
+    @InputFiles
+    @Classpath
+    final FileCollection templateFiles
 
     /**
-     * All template strings that have been set by a user.
+     * All template data that has been set by a user.
      */
-    ListProperty<String> templateStrings
+    @Input
+    final ListProperty<SerializedTemplate> templates
 
     @Inject
-    Templates(ObjectFactory objects, Project project) {
-        templateStrings = objects.listProperty(String)
+    Templates(ObjectFactory objects) {
+        classes = objects.newInstance(ClassTemplates)
+        methods = objects.newInstance(MethodTemplates)
+        entryPoints = objects.newInstance(EntryPointTemplates)
+
+        handlers = objects.listProperty(TemplateHandler).tap {
+            add(classes)
+            add(methods)
+            add(entryPoints)
+        }
 
         def handlers = this.handlers
-        templateStrings.set(
-                project.provider {
-                    handlers.collectMany { it.strings }
-                }
-        )
-
-        templateFiles = objects.listProperty(File)
-        templateFiles.set(
-                project.provider {
-                    List<File> files = handlers.collectMany { it.files }
-
-                    if (dir) {
-                        // resolve all of them if there is a directory
-                        def dirPath = dir.toPath()
-                        files = files.collect {
-                            dirPath.resolve(it.toPath()).toFile()
-                        }
-                    }
-
-                    files
-                }
-        )
-    }
+        templateFiles = objects.fileCollection()
+                .from(handlers.map { h -> h.collect { it.templateFiles } })
 
-    /**
-     * Loads all the templates from the inner configurations and reports the status as it progresses. 
-     * @param report
-     * @return All templates overridden. 
-     */
-    List<Template> getTemplates(AssertionsGeneratorReport report) {
-        // resolve user templates directory
-        if (dir == null) dir = new File(".")
+        templates = objects.listProperty(SerializedTemplate).tap {
+            addAll(classes.templates)
+            addAll(methods.templates)
+            addAll(entryPoints.templates)
+        }
+    }
 
-        // load any templates overridden by the user
-        List<Template> userTemplates = new ArrayList<>()
-        handlers.each { handler -> handler.getTemplates(userTemplates, report) }
+    static class FileOrTemplate {
+        RegularFileProperty file
+        Property<String> template
 
-        return userTemplates
+        @Inject
+        FileOrTemplate(ObjectFactory objects) {
+            file = objects.fileProperty()
+            template = objects.property(String)
+        }
     }
 
-    private void loadUserTemplate(Either<File, String> userTemplate, Template.Type type, String templateDescription,
```

Comment:
This all moved to https://github.com/assertj/assertj-generator-gradle-plugin/pull/38/files#diff-3ef36fd812e245041faaa371cb18c44d09552592692a133d1de704eb30e18aa9R37